Jimmy/7253 up sender transform constants

jimmyfagan opened the pull request:
This PR involves adding tests to verify/demonstrate that constants are appropriately handled in the `FhirTransformer`. This was pretty straightforward with the exception of the `bundleProperty` field. `bundleProperty` is parsed on our end before passing it over to `FhirPathUtils`; we need to find the path to the property being updated (including each step along that path in case we can create missing properties), as well as the property name itself, and therefore, we can't depend on the `FhirPathUtils` to handle constants as we can for the other fields which may make use of the constants. Instead, I re-introduced the `ConstantSubstitutor` so we can do some pre-processing string substitution on constants that are wrapped in `%{` and `}`.

I also stumbled upon some error cases in the process of writing these tests, so included some changes intended to make certain error cases a bit more graceful, and more clearly logged.

Test Steps:
1. Fhir Sender Transforms have still not been integrated in the pipeline, so for now, running the `FhirTransformerTests` and checking code coverage is the best way to test these changes.

## Changes
- New test cases to cover the acceptance criteria listed in #7253 in `FhirTransformerTests.kt`
- Add ConstantSubsitutor to `FhirTransformer.kt` (and update comment about ConstantSubstitutor in `FhirToHl7Converter.kt`), remove currently unneeded property `strict` from `FhirTransformer.kt`
- Adjust error handling and logging in `FhirTransformer.kt` and `FhirBundleUtils.kt`
- Add logging checks and adjust existing error testing in `FhirTransformerTests.kt`

## Checklist

### Testing
- [x] Tested locally?
- [x] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [x] Added tests?

## Linked Issues
- Fixes #8652

JessicaWNava commented on prime-router/src/main/kotlin/fhirengine/translation/hl7/FhirTransformer.kt:
```diff
@@ -172,15 +161,23 @@ class FhirTransformer(
         focusResource: Base
     ) {
-        if (bundleProperty == null) return
+        if (rawBundleProperty == null || rawBundleProperty == "") {
```
You can replace this with a check for .isNullOrBlank()

jimmyfagan replied:
Oo good call, thanks for the reminder!

JessicaWNava reviewed (approved):
One minor suggestion, other than that, looks good!

jimmyfagan commented:
Updated this with another commit to remove the support for `%{constant}` and to add some error handling to enforce some of the restrictions around the use of constants within `bundleProperty`. Ready for review again!

carlosfelix2 commented on prime-router/src/main/kotlin/fhirengine/translation/hl7/FhirTransformer.kt:
```diff
@@ -178,22 +164,42 @@ class FhirTransformer(
 
         val pathParts = bundleProperty.split(".")
+        if (pathParts.size < 2) {
+            logger.warn("Expected at least 2 parts in bundle property '$bundleProperty'.")
```
Not sure this limitation is needed. Are you sure?

jimmyfagan replied:
Yes, without it we can't use `setProperty` since it expects a property name and doesn't use fhir path.

carlosfelix2 commented on prime-router/src/main/kotlin/fhirengine/translation/hl7/FhirTransformer.kt:
```diff
@@ -213,7 +219,11 @@ class FhirTransformer(
         val property = childResource.getNamedProperty(pathParts.last())
-        val newValue = FhirBundleUtils.convertFhirType(value, value.fhirType(), property.typeCode)
+        if (property == null) {
+            logger.warn("Could not find property '${pathParts.last()}'.")
```
This may  be an error

jimmyfagan replied:
Changed to an error

carlosfelix2 commented on prime-router/src/main/kotlin/fhirengine/translation/hl7/FhirTransformer.kt:
```diff
@@ -213,7 +219,11 @@ class FhirTransformer(
-        val newValue = FhirBundleUtils.convertFhirType(value, value.fhirType(), property.typeCode)
+        if (property == null) {
+            logger.warn("Could not find property '${pathParts.last()}'.")
+            return
```
Too are many returns in this function. Is there a way to refactor this?

jimmyfagan replied:
I extracted some of the validation at the beginning of the function into a helper so there is one return there instead of three, and then reordered some logic to remove another one. Hopefully that helps. Will check that in with a new commit shortly.

carlosfelix2 commented on prime-router/src/main/kotlin/fhirengine/translation/hl7/utils/FhirBundleUtils.kt:
```diff
@@ -40,7 +41,7 @@ object FhirBundleUtils : Logging {
      * and logs an error if the conversion is not supported.
      */
-    fun convertFhirType(value: Base, sourceType: String, targetType: String): Base {
+    fun convertFhirType(value: Base, sourceType: String, targetType: String, logger: KotlinLogger = this.logger): Base {
```
logger should not be needed as an argument. It is already provided by the class

jimmyfagan replied:
It's an optional argument, by default uses the one on the class, but accepts a different logger to assist in testing the logging with mocks. From your other comments it sounds like you're not a fan of testing the logging, so let's discuss this as a 17th and we can decide how to handle this.

jimmyfagan replied:
I do think this is needed here to help with the logging mocking, even if the specific messages still aren't being verified, it's nice to be able to verify that an error or warning is logged.

carlosfelix2 commented on prime-router/src/test/kotlin/fhirengine/translation/hl7/FhirTransformerTests.kt:
```diff
@@ -332,42 +352,222 @@ class FhirTransformerTests {
+            "Bundle.entry.resource.ofType(Patient).extension(regexNonMatch).value[x]", IdType("newId"),
+            CustomContext(bundle, bundle), bundle, bundle
+        )
+        verify(exactly = 1) { logger.warn(match<String> { it.contains("Could not find property") }) }
```
Brittle tests. Remove all the tests for the specific message in the log. This also makes it sound like it should be an error.

jimmyfagan replied:
Changed to check for an error without checking for specific message.

carlosfelix2 commented on prime-router/src/test/kotlin/fhirengine/translation/hl7/FhirTransformerTests.kt:
```diff
@@ -332,42 +352,222 @@ class FhirTransformerTests {
+            "id", IdType("newId"),
+            CustomContext(bundle, bundle), bundle, bundle
+        )
+        verify(exactly = 1) { logger.warn(match<String> { it.contains("Expected at least 2 parts") }) }
```
Brittle tests. Remove all the tests for the specific message in the log. This also makes it sound like it should be an error.

jimmyfagan replied:
Changed to check for an error without checking for specific message

carlosfelix2 commented on prime-router/src/test/kotlin/fhirengine/translation/hl7/FhirTransformerTests.kt:
```diff
@@ -332,42 +352,222 @@ class FhirTransformerTests {
+            bundle, bundle
+        )
+        verify(exactly = 1) {
+            logger.warn(match<String> { it.contains("cannot dynamically create components relying on constants") })
```
Brittle tests. Remove all the tests for the specific message in the log. This also makes it sound like it should be an error.

jimmyfagan replied:
Changed to check for an error without checking for specific message, also moved some of these cases into a new test.

carlosfelix2 commented on prime-router/src/test/kotlin/fhirengine/translation/hl7/FhirTransformerTests.kt:
```diff
@@ -332,42 +352,222 @@ class FhirTransformerTests {
+            "Bundle.entry.resource.ofType(Patient).name.%key", StringType("SomeName"),
+            CustomContext(bundle, bundle, constants = mutableMapOf(Pair("key", "text"))), bundle, bundle
+        )
+        verify(exactly = 1) { logger.warn(match<String> { it.contains("Constants not supported in lowest level") }) }
```
Your test are verifying logs and not results. It makes it seem there should be some return to this functions and perhaps needs a refactor.

carlosfelix2 commented on prime-router/src/test/kotlin/fhirengine/translation/hl7/FhirTransformerTests.kt:
```diff
@@ -332,42 +352,222 @@ class FhirTransformerTests {
                 CustomContext(bundle, bundle), bundle, bundle
             )
         }.isFailure()
+        verify(exactly = 1) { logger.error(match<String> { it.contains("between CodeableConcept and string") }) }
```
Remove

jimmyfagan replied:
Done

carlosfelix2 commented on prime-router/src/test/kotlin/fhirengine/translation/hl7/FhirTransformerTests.kt:
```diff
@@ -332,42 +352,222 @@ class FhirTransformerTests {
                 CustomContext(bundle, bundle), bundle, bundle
             )
         }.isFailure()
+        verify(exactly = 1) { logger.error(match<String> { it.contains("between string and boolean") }) }
```
Remove

jimmyfagan commented:
In my latest two commits, I made some refactors to hopefully make functions easier to read. Part of this involved extracting a function for validation of inputs. I also removed the parts of tests that looked for specific logging messages and instead will just verify that errors did happen. I also changed most/all of the warnings in the function to errors. I think as part of the integration into the pipeline, there will be more error handling work to be done, but I don't want to get more into that as part of this PR, since it's not really related to this story.
